analyzers/src/SonarAnalyzer.Common/Rules/AvoidUnsealedAttributesBase.cs

```diff
@@ -0,0 +1,50 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class AvoidUnsealedAttributesBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S4060";
+
+    protected override string MessageFormat => "Seal this attribute or make it abstract.";
+
+    protected AvoidUnsealedAttributesBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context)
+        => context.RegisterSyntaxNodeActionInNonGenerated(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (Language.Syntax.NodeIdentifier(c.Node) is { } identifier
+                    && !identifier.IsMissing
+                    && c.SemanticModel.GetDeclaredSymbol(c.Node) is INamedTypeSymbol { IsAbstract: false, IsSealed: false } symbol
+                    && symbol.DerivesFrom(KnownType.System_Attribute)
+                    && symbol.IsPubliclyAccessible())
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, identifier.GetLocation()));
+                }
+            },
+            Language.SyntaxKind.ClassAndRecordDeclaration);
```

Comment:
Record cannot inherit attribute and we should not register for it. You'll need to add `ClassDeclaration` to the facade
```suggestion
            Language.SyntaxKind.ClassDeclaration);
```
```suggestion
            Language.SyntaxKind.ClassDeclaration);
```